### PR TITLE
Flex scheduler

### PR DESCRIPTION
**Pull request by yukavio**
## Motivation
Implement a better dispatch scheduler for DP mode, which could dispatch new requests depending on the remaining resources of different inference processes. It could help the server get better TTFT with high request rate compare to round-robin algorithm.

## Checklist

- [x] Before submitting a PR for review, make sure it has passed verification in your local development environment **at least**.
- [x] Ensure pre-commit `pre-commit run --all-files` or other linting tools are used to fix potential lint issues.
- [x] Confirm that modifications are covered by complete unit tests. If not, please add more unit tests for correctness.
- [x] Modify documentation as needed, such as docstrings or example tutorials.

### CODE REVIEW

**Comment by merrymercy** [comment]
Hi @yukavio, can you briefly describe the context of this PR?

**Comment by yukavio** [comment]
> Hi @yukavio, can you briefly describe the context of this PR?

I am trying to implement a load balancing strategy that is better than round robin when using Data Parallel.

**Comment by merrymercy** [comment]
If you can provide some context, high-level descriptions, and performance numbers, it can help us understand this PR better.

**Comment by yukavio** [comment]
> If you can provide some context, high-level descriptions, and performance numbers, it can help us understand this PR better.

No problem. The entire strategy is not yet fully determined and I am still trying to do further iterative optimization. I will provide an overall description and corresponding performance data after completion.

**Comment by yukavio** [comment]
We implement a resources-aware scheduler in dp mode which could be enabled with `--load-balance-method resources_aware`. We tested the performance of the resources-aware scheduler and compared it with the round-robin algorithm on an 8*A100(40G) node.
The detail of bench command: python3 -m sglang.bench_serving --backend sglang  --dataset-name random --tokenizer Qwen/Qwen2-7B --model Qwen/Qwen2-7B --random-output-len 1024 --random-input-len 4096 --random-range-ratio 0.5 --seed 1234 --num-prompts 20000 --request-rate 16.0

round_robin:
round_robin


resouces_aware:
recources_aware

and if we change the argument of num-continue-decode-step from 10 to 1 in global config, we could get better on ttft:
resources_aware2

**Comment by zhyncs** [comment]
Hi @yukavio Nice work! Could you resolve the conflicts? Thanks.

**Comment by zhyncs** [comment]
@yukavio Also, it would be better to benchmark Llama 3.1 8B Instruct and Llama 3.1 70B Instruct. Thanks.

**Comment by yukavio** [comment]
> @yukavio Also, it would be better to benchmark Llama 3.1 8B Instruct and Llama 3.1 70B Instruct. Thanks.

@zhyncs  I can provide performance comparison results on LLAMA3.1 8B later. However, for LLAMA3.1 70B, the improvements in this PR are mainly based on DP. But if it's a 70GB model, it requires multiple 8-GPU machines to start enough DP Workers to conduct this test. For me, it's difficult to gather that many machines for testing.

**Comment by zhyncs** [comment]
> But if it's a 70GB model, it requires multiple 8-GPU machines to start enough DP Workers to conduct this test.

@Ying1123 May you help take a look? Thanks.

File: python/sglang/srt/managers/controller_multi.py [hunk 1]
```diff
@@ -64,7 +70,11 @@ class WorkerHandle:
     queue: multiprocessing.Queue
 
 
-class ControllerMulti:
+# class FlexScheduler:
+#     """A scheduler which dispatch """
+
+
+class ControllerMultiFlex:
```
Comment (merrymercy) [inline python/sglang/srt/managers/controller_multi.py]: revert the name change

**Review by merrymercy — COMMENTED** [review]
please resolve the conflicts, fix the failed test cases, and add a new test case for data parallelism.

File: server.sh [hunk 2]
```diff
@@ -0,0 +1 @@
+python3 -m sglang.launch_server --model-path Qwen/Qwen1.5-1.8B-Chat  --host 0.0.0.0 --port 8080 --mem-fraction-static 0.6  --chunked-prefill-size 512
```
Comment (min-xu-et) [inline server.sh:1]: no need to check this in?

**Comment by yukavio** [comment]
> please resolve the conflicts, fix the failed test cases, and add a new test case for data parallelism.

I met some problems when merging upstream branch and I am trying to fix them. I will push the fix commit later and add a test case for it.

**Comment by yukavio** [comment]
> please resolve the conflicts, fix the failed test cases, and add a new test case for data parallelism.


We met memory problems after merging the latest main branch. Details can be found at https://github.com/sgl-project/sglang/issues/1405. It looks like the latest main branch has some issues with memory management which did not happen in the old version.

**Comment by merrymercy** [comment]
https://github.com/sgl-project/sglang/issues/1405#issuecomment-2366743378

File: python/sglang/srt/managers/controller_multi.py [hunk 3]
```diff
@@ -135,6 +148,41 @@ def start_dp_worker(self, dp_worker_id: int):
...
+                    i for i, x in enumerate(num_reqs_waiting) if x == min_value
+                ]
+
+                # find max available mem from waiting queue
+                index = max(min_indices, key=lambda i: available_mem[i])
+                self.workers[index].queue.put(r)
+                num_reqs_waiting[index] += 1
+                available_mem[index] -= len(r.input_ids)
+            else:
+                # find no waitting and max available mem node
```
Comment (daquexian) [inline python/sglang/srt/managers/controller_multi.py]: nit: waitting -> waiting

**Comment by yukavio** [comment]
The code of this PR was reformed by my teammates who created a new PR https://github.com/sgl-project/sglang/pull/1913. I will close this PR.

**Pull request closed without merging**